src/main.rs

```diff
@@ -1,3 +1,207 @@
+
+fn main() -> io::Result<()> {
+    let cli = Cli::parse();
+
+    if let Some(jobs) = cli.jobs {
+        rayon::ThreadPoolBuilder::new().num_threads(jobs).build_global().unwrap();
+    }
+
+    let content = fs::read_to_string(cli.file)?;
+
+    let chat: Chat = serde_json::from_str(&content).unwrap();
+    let stat: ChatStatistics = ChatStatistics::gather(&chat);
+
+    let file = fs::File::create(cli.output)?;
+    serde_json::to_writer_pretty(file, &stat)?;
+
+    Ok(())
+}
+
+#[derive(Debug, Parser)]
+struct Cli {
+    #[arg(long, short)]
+    file: PathBuf,
+    #[arg(long, short, default_value = "out.json")]
+    output: PathBuf,
+    #[arg(long, short)]
+    jobs: Option<usize>
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Person<'a>(Cow<'a, str>);
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq, Hash, Clone)]
+struct Token<'a>(Cow<'a, str>);
+
+impl<'a> From<&'a str> for Token<'a> {
+    fn from(value: &'a str) -> Self {
+        Token(value.into())
+    }
+}
+impl<'a> From<String> for Token<'a> {
+    fn from(value: String) -> Self {
+        Token(value.into())
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Chat<'a> {
+    name: String,
+    #[serde(rename = "type")]
+    chat_type: ChatType,
+    id: u128,
+    messages: Vec<Message<'a>>,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+enum ChatType {
+    PublicChannel,
+    PrivateChannel,
+    PublicSupergroup,
+    PrivateSupergroup,
+    PersonalChat,
+    ChatForbidden,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Message<'a> {
+    id: u64,
+    #[serde(rename = "type")]
+    message_type: MessageType,
+    date: NaiveDateTime,
+    from: Option<Person<'a>>,
+    text_entities: Vec<TextEntity>,
+}
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Eq)]
+#[serde(rename_all = "lowercase")]
+enum MessageType {
+    Service,
+    Message,
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TextEntity {
+    #[serde(rename = "type")]
+    text_type: TextEntityType,
+    text: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(rename_all = "snake_case")]
+enum TextEntityType {
+    Pre,
+    Bold,
+    Link,
+    Code,
+    Email,
+    Plain,
+    Phone,
+    Italic,
+    Cashtag,
+    Spoiler,
+    Mention,
+    Hashtag,
+    TextLink,
+    Underline,
+    BotCommand,
+    CustomEmoji,
+    MentionName,
+    Strikethrough,
+}
+
+impl TextEntityType {
+    pub fn is_meta(&self) -> bool {
+        use TextEntityType::*;
+        matches!(self, Phone | BotCommand | Email | CustomEmoji | Mention)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct ChatStatistics<'a> {
+    num_tokens: usize,
+    #[serde(borrow)]
+    members_tokens_map: HashMap<Person<'a>, HashMap<Token<'a>, usize>>,
+    #[serde(borrow)]
+    tokens_map: HashMap<Token<'a>, usize>,
+}
+impl<'a> ChatStatistics<'a> {
+    pub fn gather(chat: &'a Chat) -> Self {
```

Comment:
This method contains a lot of `unwrap()`. Wouldn't be better to change the return type to `Result` and use elvis operator instead of unwrapping?

Reply:
Only three times 🤷
Function is "tuned" in a way that options would always contain `Some(_)`, so no panicking is possible